torch/distributed/_composable/fsdp/_fsdp_param_group.py

```diff
@@ -166,17 +166,18 @@ def _init_grad_divide_factors(self):
         data_parallel_world_size *= self.mesh_info.shard_mesh_size
         if isinstance(self.mesh_info, HSDPMeshInfo):
             data_parallel_world_size *= self.mesh_info.replicate_mesh_size
-        if self._reduce_dtype == torch.float32:
+        if self._reduce_dtype in (torch.float32, torch.bfloat16):
             # Use NCCL's AVG op to divide after reduction since it is more
             # performant and fp32 has sufficient precision
             self._grad_divide_factors: Union[Tuple[None, None], Tuple[float, float]] = (
                 None,
                 None,
             )
             return
-        # For N data parallel workers, each worker computes g_i, and they
-        # collectively reduce (g_1 + ... + g_N) / N. To avoid overflow and
-        # underflow, we divide by ~sqrt(N) before and after the reduction.
+        # Since fp16 has smaller dynamic range than fp32/bf16, we want to avoid
+        # overflow/underflow. For N data parallel workers, each worker computes
+        # g_i, and they collectively reduce (g_1 + ... + g_N) / N. To avoid
+        # overflow/underflow, we divide by ~sqrt(N) before/after the reduction.
```

Comment:
> To avoid overflow/underflow, we divide by ~sqrt(N) before/after the reduction

Not directly related to this PR, but even in this case (fp16), we can still save the "divide by ~sqrt(N) before the reduction" by using the nccl PreMulSum reduce op:

https://docs.nvidia.com/deeplearning/nccl/archives/nccl_21212/user-guide/docs/api/ops.html

You give this op a `1 / sqrt(N)` scaling factor.

Reply:
IIUC, we still need to manually divide by the `1 / sqrt(N)` in that case because we do not have a "PostMulSum"?

Reply:
Yeah, you would still need to do a manual post division. Just saving the pre division.